Philip McCormick

MSDS 599

April 4, 2022

Homework 2

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
from sklearn.model_selection import cross_validate
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn import datasets
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier

In [4]:
df=pd.read_csv('/content/drive/MyDrive/Data 599/carseats.csv')
df.head()

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0   9.50        138      73           11         276    120       Bad   42   
1  11.22        111      48           16         260     83      Good   65   
2  10.06        113      35           10         269     80    Medium   59   
3   7.40        117     100            4         466     97    Medium   55   
4   4.15        141      64            3         340    128       Bad   38   

   Education Urban   US  
0         17   Yes  Yes  
1         10   Yes  Yes  
2         12   Yes  Yes  
3         14   Yes  Yes  
4         13   Yes   No

In [5]:
df.dtypes

Sales          float64
CompPrice        int64
Income           int64
Advertising      int64
Population       int64
Price            int64
ShelveLoc       object
Age              int64
Education        int64
Urban           object
US              object
dtype: object

1. Add a column called 'Class' that is 1 if Sales >=8 and 0 otherwise. (Be sure to drop the 'Sales' column once you're done!) 2


In [6]:
df.isna().sum()

Sales          0
CompPrice      0
Income         0
Advertising    0
Population     0
Price          0
ShelveLoc      0
Age            0
Education      0
Urban          0
US             0
dtype: int64

In [7]:
df['Class'] = np.where(df['Sales'] >= 8, 1, 0)
df.head()

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0   9.50        138      73           11         276    120       Bad   42   
1  11.22        111      48           16         260     83      Good   65   
2  10.06        113      35           10         269     80    Medium   59   
3   7.40        117     100            4         466     97    Medium   55   
4   4.15        141      64            3         340    128       Bad   38   

   Education Urban   US  Class  
0         17   Yes  Yes      1  
1         10   Yes  Yes      1  
2         12   Yes  Yes      1  
3         14   Yes  Yes      0  
4         13   Yes   No      0

In [8]:
df.dtypes

Sales          float64
CompPrice        int64
Income           int64
Advertising      int64
Population       int64
Price            int64
ShelveLoc       object
Age              int64
Education        int64
Urban           object
US              object
Class            int64
dtype: object

In [9]:
df.drop(columns=['Sales'])

CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0          138      73           11         276    120       Bad   42   
1          111      48           16         260     83      Good   65   
2          113      35           10         269     80    Medium   59   
3          117     100            4         466     97    Medium   55   
4          141      64            3         340    128       Bad   38   
..         ...     ...          ...         ...    ...       ...  ...   
395        138     108           17         203    128      Good   33   
396        139      23            3          37    120    Medium   55   
397        162      26           12         368    159    Medium   40   
398        100      79            7         284     95       Bad   50   
399        134      37            0          27    120      Good   49   

     Education Urban   US  Class  
0           17   Yes  Yes      1  
1           10   Yes  Yes      1  
2           12   Yes  Yes      1  
3           14   Yes  Yes      0  
4           13   Yes   No      0  
..         ...   ...  ...    ...  
395         14   Yes  Yes      1  
396         11    No  Yes      0  
397         18   Yes  Yes      0  
398         12   Yes  Yes      0  
399         16   Yes  Yes      1  

[400 rows x 11 columns]

2. One-hot encode the categorical variables (see the pd.get_dummies() function). 2

In [10]:
df = pd.get_dummies(df)

3. Split the data set into training and test sets. Use test_size=0.3, random_state=5 in the train_test_split call. 2

In [11]:
x=df.drop('Class',axis=1)
y=df['Class']

In [12]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=5)

4. Scale the data using StandardScaler() 2

In [13]:
from sklearn.preprocessing import StandardScaler

In [14]:
scaler=StandardScaler()
x_train_scaled=scaler.fit_transform(x_train)
x_test_scaled=scaler.fit_transform(x_test)

In [15]:
np.allclose(np.mean(x_train_scaled,axis=0),0),np.allclose(np.std(x_train_scaled,axis=0),1)

(True, True)

5. Use sklearn to fit a logistic regression model on the training set and find the accuracy on the test set. 5

In [16]:
car_class = LogisticRegression(random_state=0,penalty='none').fit(x_train_scaled, y_train)

In [17]:
print(car_class.coef_,car_class.intercept_)

[[ 3.84707004e+02  7.24358019e+00 -1.06755525e+01  1.04333321e+01
   5.42359631e+00 -1.70279379e+01 -6.20075594e+00 -5.75968241e-02
  -3.88672122e+00  8.44443558e+00 -3.72687741e+00  3.91516918e+00
  -3.91516918e+00 -2.97831849e-01  2.97831849e-01]] [-88.18207011]


In [25]:
accuracy_score(car_class.predict(x_test_scaled),y_test)

0.95

6. Now use k nearest neighbors to build a classifier. Use the same training and test sets as above. Use cross validation to choose the optimal k. 5

In [19]:
cvres=[]
for i in range(1,100):
    clf = KNeighborsClassifier(n_neighbors=i)
    nn = cross_validate(clf, x_train_scaled, y_train, cv=5, scoring='accuracy')
    cvres.append(nn['test_score'].mean())

In [20]:
nn['test_score']

array([0.76785714, 0.75      , 0.82142857, 0.75      , 0.78571429])

In [21]:
np.argmax(cvres)

12

In [26]:
cvres[0:15]

[0.7964285714285714,
 0.7857142857142857,
 0.8285714285714286,
 0.7892857142857144,
 0.8107142857142857,
 0.8,
 0.8071428571428572,
 0.7892857142857143,
 0.8178571428571428,
 0.8,
 0.8285714285714285,
 0.8321428571428571,
 0.8392857142857142,
 0.8392857142857142,
 0.8392857142857142]

In [23]:
mod = KNeighborsClassifier(n_neighbors=13)
nb12=mod.fit(x_train_scaled,y_train)

In [24]:
accuracy_score(nb12.predict(x_test_scaled),y_test)

0.7583333333333333

7. Comment on your results. 2

The train model fits the test model with 75.8% accuracy. This is considered a relatively high accuracy, but depending on the context of what we are trying to classify it may not be sufficient. For looking at car sales this may be useful but if we were to be interpretting health outcomes this percentage of accuracy my be a matter of life and death for patients and could be deemed an unacceptable accuracy.